### DISCUSSION TITLE
Are view collections materialized?

### DISCUSSION THREAD

**Discussion by tero-kiituri** [pocketbase/pocketbase#3836] [Q&A] [answered]
Hi! First of all thanks a lot for the Pocketbase. Awesome work!

I'm wondering if the view collections are materialized or are view queries run every time the view is queried?

**Comment by ganigeorgiev** [comment, marked as answer]
View collections are backed by the SQLite views and there is no support at the moment for materialized views, aka. there is no separate persistence for the view data and view queries are executed every time when the view collection is queried.

The performance of a view query is expected to be roughly the same as if executing custom db query against your collection(s) (_the SQLite query planner will try to use the underlying collection tables indexes when applicable_).

**Reply by tero-kiituri** [reply]
Ok. Thanks for the answer!

There's something funny going on with my query and I can't figure out what. (Maybe because I don't understand SQL and SQLite well enough.) There is about 6M rows in the biggest table and all the tables are joined together and then the result set is filtered with where clause. First I thought that 8 joins in the query and/or the dataset's size are too much and it's slow because of them. That's why I asked the first question.

The search/query takes something like 30 to 40 mississippies to execute in PB admin UI and results "context deadline exceeded" to be printed into log. Nevertheless I get the result back to the UI. Execution of an equivalent query in SQLite console takes only 20 to 25 mississippies.

However, now it seems that, the joins and the size of the dataset are neither the problem. It's the query's where clause. There are comparison of two different name fields (both text type), a comparison of a date field (apparently text type too), and AND operations between them. If I remove the date condition the query executes nearly instantaneously even if it results larger data set. This happens both in the PB admin UI and in the SQLite console. Order of the comparisons doesn't seem to matter.

Do you happen to know if there is something special to take into notice with SQLite and date fields?

**Reply by ganigeorgiev** [reply]
SQLite doesn't have a dedicated `date` type and we use just `TEXT` to store the value.

About the difference between the Admin UI execution and the one in the SQLite cli - if you use the prebuilt executable note that by default PocketBase uses a pure Go port of the C SQLite code and it is expected to be slightly slower in some situations. If you want a more close to native performance you can build with `CGO_ENABLED=1` but note that this will complicate the cross-compilation as the binary will no longer be statically linked. PocketBase also perform serialization, access checks, etc. so this could also affect the retrieval performance.

In any case, the best way to find why a specific condition affect so much your query performance is to inspect what the query planner do using `EXPLAIN QUERY PLAN yourquery` in the sqlite3 cli.

**Reply by ganigeorgiev** [reply]
If you are not able to identify the culprit, feel free to paste your collections configuration as json (Admin UI > Settings > Export collections) and I'll try to find time to investigate it.

**Reply by tero-kiituri** [reply]
I created some indexes straight to the db and that solved the problem. All the queries are now blazingly fast.

Is there way to add index creations to migration files?

**Reply by tero-kiituri** [reply]
I answer my own question... PB in framework mode makes possible to add custom queries (https://pocketbase.io/docs/go-database/) and with those you can create custom e.g. custom indexes.

```
res, err := app.Dao().DB().
    NewQuery("CREATE INDEX name_idx ON users (name)").
    Execute()
```

And if I think this right the migration files would indeed be the best place for these custom queries.

**Reply by ganigeorgiev** [reply]
If you need to add indexes to a collection, prefer adding them through the collection model (eg. https://pocketbase.io/docs/js-collections/#create-new-collection-without-data-validations).

Indexes can be also added from the Admin UI > Edit collection > "Unique constraints and indexes". This will generate the migration file for you.

**Reply by tero-kiituri** [reply]
Oh, nice! I didn't even realize it's possible to do them from the admin UI. That's probably the best option. Thanks! 👍
